builtins/web/form-data/form-data-encoder.cpp

```diff
@@ -0,0 +1,492 @@
+        len += newline_len;
+        i++;
+      } else {
+        len += newline_len;
+      }
+    } else if (src[i] == LF) {
+      len += newline_len;
+    } else {
+      len += 1;
+    }
+  }
+
+  return len;
+}
+
+// Replace every occurrence of U+000D (CR) not followed by U+000A (LF),
+// and every occurrence of U+000A (LF) not preceded by U+000D (CR),
+// in entry's name, by a string consisting of a U+000D (CR) and U+000A (LF).
+//
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#multipart-form-data
+std::optional<std::string> normalize_newlines(std::string_view src) {
+  std::string output;
+
+  output.reserve(compute_normalized_len(src, CRLF));
+
+  for (size_t i = 0; i < src.size(); i++) {
+    if (src[i] == CR) {
+      if (i + 1 < src.size() && src[i + 1] == LF) {
+        output += CRLF;
+        i++;
+      } else {
+        output += CRLF;
+      }
+    } else if (src[i] == LF) {
+      output += CRLF;
+    } else {
+      output.push_back(src[i]);
+    }
+  }
+
+  return output;
+}
+
+std::optional<std::string> normalize_newlines(JSContext *cx, HandleValue src) {
+  auto chars = core::encode(cx, src);
+  if (!chars) {
+    return std::nullopt;
+  }
+
+  return normalize_newlines(chars);
+}
+
+// For field names and filenames for file fields, the result of the encoding must
+// be escaped by replacing any 0x0A (LF) bytes with the byte sequence
+// `%0A`, 0x0D (CR) with `%0D` and 0x22 (") with `%22`.
+//
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#multipart-form-data
+std::optional<std::string> escape_newlines(std::string_view str) {
+  int32_t offset = 0;
+  std::string output(str);
+
+  while ((offset = output.find_first_of("\n\r\"", offset)) != std::string::npos) {
+    if (output[offset] == '\n') {
+      output.replace(offset, 1, "%0A");
+    } else if (output[offset] == '\r') {
+      output.replace(offset, 1, "%0D");
+    } else if (output[offset] == '"') {
+      output.replace(offset, 1, "%22");
+    } else {
+      offset++;
+      continue;
+    }
+  }
+
+  return output;
+}
+
+std::optional<std::string> escape_newlines(JSContext *cx, HandleValue src) {
+  auto chars = core::encode(cx, src);
+  if (!chars) {
+    return std::nullopt;
+  }
+
+  return escape_newlines(chars);
+}
+
+} // namespace
+
+namespace builtins {
+namespace web {
+namespace form_data {
+
+using blob::Blob;
+using file::File;
+using streams::BufReader;
+using streams::NativeStreamSource;
+
+using EntryList = JS::GCVector<FormDataEntry, 0, js::SystemAllocPolicy>;
+
+struct StreamContext {
+  StreamContext(const EntryList *entries, std::span<uint8_t> outbuf)
+      : entries(entries), outbuf(outbuf), read(0), done(false) {}
+  const EntryList *entries;
+
+  std::span<uint8_t> outbuf;
+  size_t read;
+  bool done;
+
+  size_t remaining() {
+    MOZ_ASSERT(outbuf.size() >= read);
+    return outbuf.size() - read;
+  }
+
+  template <typename I> size_t write(I first, I last) {
```

Comment:
IIUC, this method is deliberately infallible? If so, could you document that fact?